// @shopify/hydrogen/dist/node/framework/Hydration/ServerComponentResponse.server.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerComponentResponse = void 0;
const CachingStrategy_1 = require("../CachingStrategy");
const Redirect_client_1 = __importDefault(require("../../foundation/Redirect/Redirect.client"));
const react_1 = __importDefault(require("react"));
class ServerComponentResponse extends Response {
    constructor() {
        super(...arguments);
        this.wait = false;
        this.cacheOptions = (0, CachingStrategy_1.CacheSeconds)();
    }
    /**
     * Buffer the current response until all queries have resolved,
     * and prevent it from streaming back early.
     */
    doNotStream() {
        this.wait = true;
    }
    canStream() {
        return !this.wait;
    }
    cache(options) {
        this.cacheOptions = options;
    }
    get cacheControlHeader() {
        return (0, CachingStrategy_1.generateCacheControlHeader)(this.cacheOptions);
    }
    writeHead({ status, statusText, headers, } = {}) {
        if (status || statusText) {
            this.customStatus = { code: status, text: statusText };
        }
        if (headers) {
            for (const [key, value] of Object.entries(headers)) {
                this.headers.set(key, value);
            }
        }
    }
    redirect(location, status = 307) {
        // writeHead is used for SSR, so that the server responds with a redirect
        this.writeHead({ status, headers: { location } });
        // in the case of an RSC request, instead render a client component that will redirect
        return react_1.default.createElement(Redirect_client_1.default, { to: location });
    }
}
exports.ServerComponentResponse = ServerComponentResponse;
