// @shopify/hydrogen/dist/node/foundation/Analytics/Analytics.client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Analytics = void 0;
const react_1 = require("react");
const ClientAnalytics_1 = require("./ClientAnalytics");
function Analytics({ analyticsDataFromServer, }) {
    (0, react_1.useEffect)(() => {
        const urlParams = new URLSearchParams(window.location.search);
        if (urlParams.has('utm_source')) {
            ClientAnalytics_1.ClientAnalytics.pushToPageAnalyticsData({
                id: urlParams.get('utm_id'),
                source: urlParams.get('utm_source'),
                campaign: urlParams.get('utm_campaign'),
                medium: urlParams.get('utm_medium'),
                content: urlParams.get('utm_content'),
                term: urlParams.get('utm_term'),
            }, 'utm');
        }
        ClientAnalytics_1.ClientAnalytics.pushToPageAnalyticsData(analyticsDataFromServer);
        ClientAnalytics_1.ClientAnalytics.publish(ClientAnalytics_1.ClientAnalytics.eventNames.PAGE_VIEW, true);
        if (analyticsDataFromServer.publishEventsOnNavigate) {
            analyticsDataFromServer.publishEventsOnNavigate.forEach((eventName) => {
                ClientAnalytics_1.ClientAnalytics.publish(eventName, true);
            });
        }
        return function cleanup() {
            ClientAnalytics_1.ClientAnalytics.resetPageAnalyticsData();
        };
    }, [analyticsDataFromServer]);
    return null;
}
exports.Analytics = Analytics;
