// @shopify/plugin-ngrok/dist/tunnel.js
import { error, os, output, ui } from '@shopify/cli-kit';
import ngrok from 'ngrok';
const NgrokError = (ngrokErrorMessage) => {
    return new error.Abort(`The ngrok tunnel could not be started.\n\n${ngrokErrorMessage}`, buildTryMessage(ngrokErrorMessage));
};
export async function start(options) {
    if (!(await ngrok.validConfig())) {
        const token = await tokenPrompt();
        await authenticate(token);
    }
    return ngrok.connect({ proto: 'http', addr: options.port }).catch((e) => {
        throw NgrokError(e.message);
    });
}
export async function authenticate(token) {
    if (!token) {
        token = await tokenPrompt(false);
    }
    await ngrok.authtoken(token);
    await ngrok.upgradeConfig();
}
async function tokenPrompt(showExplanation = true) {
    const explanation = showExplanation ? '\nTo make your local code accessible to your dev store, you need to use a ' +
        'Shopify-trusted tunneling service called ngrok. ' : '';
    const ngrokURL = 'https://dashboard.ngrok.com/get-started/your-authtoken';
    const link = output.token.link(ngrokURL, ngrokURL);
    output.info(output.content `${explanation}To sign up and get an auth token: ${link}\n`);
    const question = {
        type: 'password',
        name: 'token',
        message: 'Enter your ngrok token.',
        validate: (value) => {
            if (value.length === 0) {
                return 'Token cannot be empty';
            }
            return true;
        },
    };
    const input = await ui.prompt([question]);
    return input.token;
}
function buildTryMessage(nrokErrorMessage) {
    if (/err_ngrok_108/.test(nrokErrorMessage)) {
        const { platform } = os.platformAndArch();
        let tryMessage = 'Kill all the ngrok processes with ';
        if (platform === 'windows') {
            return tryMessage.concat(output.content `${output.token.genericShellCommand('taskkill /f /im ngrok.exe')}`.value);
        }
        else {
            return tryMessage.concat(output.content `${output.token.genericShellCommand('killall ngrok')}`.value);
        }
    }
    else if (/err_ngrok_105|err_ngrok_106|err_ngrok_107/.test(nrokErrorMessage)) {
        return output.content `Update your ngrok token with ${output.token.genericShellCommand('shopify ngrok auth')}`.value;
    }
    return undefined;
}
